### PR TITLE
ENH: Add 3d cone cude performance example with ASTRA

### PR DESCRIPTION
**Pull request by adler-j**
Minor example. Clearly needs another example where we use pygpu on the ODL side.

### CODE REVIEW

File: examples/tomo/backends/astra_performance_cuda_cone_3d_cg.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+# Solve with CGLS (aka CGN)
+x = reco_space.zero()
+with odl.util.Timer('ODL run'):
+    odl.solvers.conjugate_gradient_normal(ray_trafo, x, rhs, niter=niter)
+
+coords = (slice(None), slice(None), 128)
+
+# Display results for comparison
+plt.figure('data')
```
Comment (kohr-h) [inline examples/tomo/backends/astra_performance_cuda_cone_3d_cg.py]: wrong title (or wrong input to plot)

**Comment by adler-j** [comment]
My runtime diff is about the same as you with astra but 3 times that with ODL, try running the profiler on your machine and see what's wrong.

**Comment by kohr-h** [comment]
Here's the culprit:

profile

It's the call in [this line](https://github.com/odlgroup/odl/blob/05873fdc6e1bd8769135608ce871ef46b19cf8a4/odl/tomo/backends/astra_cuda.py#L274-L275).

**Comment by adler-j** [comment]
I have nothing like that on my machine, weird. Can you try to fix it on yours?

**Comment by kohr-h** [comment]
Same runtime in the console, so no issue with Spyder.

**Comment by kohr-h** [comment]
Same result in a fresh conda environment.

No idea what's going on.

**Comment by kohr-h** [comment]
Okay, I had a suspicion and it was correct. There's some disk swapping going on in my home directory, which is on a network drive. Therefore it's bloody slow.

**Comment by adler-j** [comment]
How much RAM do you have? Does the problem not fit in your RAM?

**Comment by kohr-h** [comment]
> How much RAM do you have? Does the problem not fit in your RAM?

I've got 64G so that shouldn't be a problem. Also swap is on a hard disk somewhere else, so it's not really swapping. But when I monitor the the disk usage I see changes in my home dir (on a network FS) during the run.

**Comment by adler-j** [comment]
Does ASTRA access the disk at any point?

**Comment by kohr-h** [comment]
I don't think so. The pure ASTRA code is roughly as fast as it should be.

**Comment by wjp** [comment]
I get roughly identical numbers as @kohr-h on a machine at CWI without a networked home directory. Using conda with odl-0.5.3-py35_2 and astra-toolbox-1.8-np111py35_1 .

**Comment by adler-j** [comment]
Very weird stuff going on there. I get

```
                     ASTRA run :      9.586 
                       ODL run :     30.476 
```

where for me `ascontiguousarray` takes a total of 9 seconds. Could you expand it further so we see if there are any weird calls under the hood? And can you run with `ODL 0.5.4.dev0`?

Anyway, it seems the call to `ascontiguousarray` is not needed and removing it I get a slight speedup

```
                     ASTRA run :      9.510 
                       ODL run :     27.949 
```

I pushed that up here, see how it goes for you now.

**Comment by adler-j** [comment]
Bump for input

**Comment by wjp** [comment]
With ODL from this branch, on an i7-3930K + GTX1080, I get timings around

```
                     ASTRA run :      5.181 
                       ODL run :     32.031 
```

However, this was not triggered by your latest commit here, but by some other commit between 0.5.3 and this branch. (I can bisect if that would be interesting or useful, but haven't yet.)

**Comment by adler-j** [comment]
Well bisecting is not useful, but it would be nice to see if the latest commit helped, anyway that number is at least better than the one before.

**Comment by kohr-h** [comment]
I'm currently working on other stuff, but I'll also rerun the example. Still getting the weird oscillations between 30 and 85 seconds @wjp?

**Comment by wjp** [comment]
For context: I saw the oscillations kohr-h is referring to with odl 0.5.3. Usually I got timings around 85 seconds, but maybe 20% of the time it was around 35 seconds instead.

I didn't see any large changes in timings caused by c1e6a4d72471cccfafdb8b613bb5b61d0c5e297f, but I'll re-run the script a few times to get some accurate numbers.

**Comment by adler-j** [comment]
Extremely weird. Can you do some memory profiling and see if you occationally run into swapping? I.e. 30 sec could be when you run on the RAM, while 85 may be when you have to run on the disk.

For reference I've tried this on both linux and windows and I'm not seeing it. My setups are:

* Windows 7, Titan X GPU
* Ubuntu 14.04, GTX 970 GPU

**Comment by wjp** [comment]
I just re-ran this script a few times with 3 versions of ODL.

There's no swapping (or any other noticable disk access).

3d23aad : 32.123 (average of 5 runs)
c1e6a4d: 32.322 (average of 5 runs)
ee326c5 (v5.3.0):  84.750 (average of 3 runs)

**Comment by adler-j** [comment]
Thank you for the input!

It would seem that 32 seconds is at least much better than 85 seconds. So perhaps that is actual overhead and i guess we can live with it until the CUDA level integration is added.

Ok to merge?

**Comment by kohr-h** [comment]
I guess we'll never know :-) 
Merging is fine with me.

**Comment by kohr-h** [comment]
Thanks @wjp for exploring!

**Comment by adler-j** [comment]
Agree, thanks @wjp for the help!

**Pull request merged**